File: backend/internal/models/unibox.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ──────────────────────────────────────────────────────────────────────────────
// Conversation
// ──────────────────────────────────────────────────────────────────────────────

type Conversation struct {
	ID                  uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	LinkedInAccountID   uuid.UUID `gorm:"column:linkedin_account_id;type:uuid;not null;index" json:"linkedin_account_id"`
	ParticipantName     string    `gorm:"type:varchar(255);not null" json:"participant_name"`
	ParticipantProfileURL *string `gorm:"type:varchar(500)" json:"participant_profile_url"`
	ParticipantHeadline *string   `gorm:"type:varchar(500)" json:"participant_headline"`
	ParticipantAvatarURL *string  `gorm:"type:text" json:"participant_avatar_url"`
	LastMessageAt       *time.Time `gorm:"type:timestamptz" json:"last_message_at"`
	LastMessagePreview  *string   `gorm:"type:text" json:"last_message_preview"`
	UnreadCount         int       `gorm:"type:integer;default:0" json:"unread_count"`
	IsArchived          bool      `gorm:"type:boolean;default:false" json:"is_archived"`
	ThreadID            string    `gorm:"type:varchar(255);uniqueIndex" json:"thread_id"`
	Label               *string   `gorm:"type:text" json:"label"`
	CreatedAt           time.Time `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt           time.Time `gorm:"autoUpdateTime" json:"updated_at"`

	// Relations
	LinkedInAccount *LinkedInAccount `gorm:"foreignKey:LinkedInAccountID;references:ID" json:"linkedin_account,omitempty"`
}

func (Conversation) TableName() string { return "conversations" }

func (c *Conversation) BeforeCreate(tx *gorm.DB) error {
	if c.ID == uuid.Nil {
		c.ID = uuid.New()
	}
	return nil
}

// ──────────────────────────────────────────────────────────────────────────────
// Message
// ──────────────────────────────────────────────────────────────────────────────

type Message struct {
	ID                uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	ConversationID    uuid.UUID `gorm:"type:uuid;not null;index" json:"conversation_id"`
	LinkedInAccountID uuid.UUID `gorm:"column:linkedin_account_id;type:uuid;not null;index" json:"linkedin_account_id"`
	MessageID         string    `gorm:"type:varchar(255);uniqueIndex" json:"message_id"`
	SenderName        string    `gorm:"type:varchar(255)" json:"sender_name"`
	SenderProfileURL  *string   `gorm:"type:varchar(500)" json:"sender_profile_url"`
	IsFromMe          bool      `gorm:"type:boolean;default:false" json:"is_from_me"`
	Content           string    `gorm:"type:text" json:"content"`
	SentAt            time.Time `gorm:"type:timestamptz;not null" json:"sent_at"`
	IsRead            bool      `gorm:"type:boolean;default:false" json:"is_read"`
	HasAttachment     bool      `gorm:"type:boolean;default:false" json:"has_attachment"`
	AttachmentURL     *string   `gorm:"type:text" json:"attachment_url"`
	CreatedAt         time.Time `gorm:"autoCreateTime" json:"created_at"`

	// Relations
	Conversation *Conversation `gorm:"foreignKey:ConversationID;references:ID" json:"-"`
}

func (Message) TableName() string { return "messages" }

func (m *Message) BeforeCreate(tx *gorm.DB) error {
	if m.ID == uuid.Nil {
		m.ID = uuid.New()
	}
	return nil
}

// ──────────────────────────────────────────────────────────────────────────────
// DTOs — Conversations
// ──────────────────────────────────────────────────────────────────────────────

// GetConversationsFilter maps to the TS filters for getConversations.
type GetConversationsFilter struct {
	AccountID   string `query:"account_id"`
	ShowArchived *bool  `query:"show_archived"` // pointer so we can distinguish missing vs false
	OnlyUnread  bool   `query:"only_unread"`
	SearchQuery string `query:"search_query"`
	Label       string `query:"label"`
}

// ArchiveConversationRequest is the body for archive/unarchive.
type ArchiveConversationRequest struct {
	Archive *bool `json:"archive"` // pointer; defaults to true when nil
}

// SetLabelRequest is the body for setting a conversation label.
type SetLabelRequest struct {
	Label *string `json:"label"` // null clears the label
}

// ──────────────────────────────────────────────────────────────────────────────
// DTOs — Messages
// ──────────────────────────────────────────────────────────────────────────────

// SendMessageRequest is the body for sending a reply.
type SendMessageRequest struct {
	ConversationID    string `json:"conversation_id" validate:"required"`
	LinkedInAccountID string `json:"linkedin_account_id" validate:"required"`
	Content           string `json:"content" validate:"required"`
}

// ──────────────────────────────────────────────────────────────────────────────
// DTOs — Sync
// ──────────────────────────────────────────────────────────────────────────────

// TriggerSyncResponse is returned after a sync trigger.
type TriggerSyncResponse struct {
	Success bool   `json:"success"`
	Error   string `json:"error,omitempty"`
}

// ──────────────────────────────────────────────────────────────────────────────
// DTOs — Campaign Context
// ──────────────────────────────────────────────────────────────────────────────

// CampaignContextResponse is inline context shown in thread view.
type CampaignContextResponse struct {
	CampaignName *string `json:"campaignName"`
	LeadStatus   *string `json:"leadStatus"`
	Company      *string `json:"company"`
	Position     *string `json:"position"`
}

File: frontend/src/types/unibox.ts
// ─── Unibox Types ────────────────────────────────────────────────────────────

// ─── Models ──────────────────────────────────────────────────────────────────

export interface Conversation {
  id: string
  linkedin_account_id: string
  participant_name: string
  participant_profile_url: string | null
  participant_headline: string | null
  participant_avatar_url: string | null
  last_message_at: string | null
  last_message_preview: string | null
  unread_count: number
  is_archived: boolean
  thread_id: string
  label: string | null
  created_at: string
  updated_at: string
  linkedin_account?: {
    id: string
    email: string
  }
}

export interface Message {
  id: string
  conversation_id: string
  linkedin_account_id: string
  message_id: string
  sender_name: string
  sender_profile_url: string | null
  is_from_me: boolean
  content: string
  sent_at: string
  is_read: boolean
  has_attachment: boolean
  attachment_url: string | null
  created_at: string
}

// ─── DTOs ────────────────────────────────────────────────────────────────────

export interface GetConversationsFilter {
  account_id?: string
  show_archived?: boolean
  only_unread?: boolean
  search_query?: string
  label?: string
}

export interface SendMessageRequest {
  conversation_id: string
  linkedin_account_id: string
  content: string
}

export interface ArchiveConversationRequest {
  archive: boolean
}

export interface SetLabelRequest {
  label: string | null
}

export interface TriggerSyncResponse {
  success: boolean
  error?: string
}

export interface CampaignContextResponse {
  campaignName: string | null
  leadStatus: string | null
  company: string | null
  position: string | null
}

// ─── Label Config ────────────────────────────────────────────────────────────

export interface LabelConfig {
  emoji: string
  color: string
  bg: string
}

export const LABELS: Record<string, LabelConfig> = {
  interested:     { emoji: '🟢', color: 'text-green-700', bg: 'bg-green-50 border-green-200' },
  maybe:          { emoji: '🟡', color: 'text-yellow-700', bg: 'bg-yellow-50 border-yellow-200' },
  not_interested: { emoji: '🔴', color: 'text-red-700', bg: 'bg-red-50 border-red-200' },
  meeting_booked: { emoji: '📅', color: 'text-blue-700', bg: 'bg-blue-50 border-blue-200' },
}

// ─── Filter Tab Type ─────────────────────────────────────────────────────────

export type ConversationFilter = 'all' | 'unread' | 'needs_reply' | 'archived'
